test/common/http/http1/codec_impl_test.cc

```diff
@@ -1913,6 +1917,76 @@ TEST_F(Http1ServerConnectionImplTest, TestSmugglingAllowChunkedContentLength100)
   testServerAllowChunkedContentLength(100, true);
 }
 
+TEST_F(Http1ServerConnectionImplTest,
+       ShouldDumpParsedAndPartialHeadersWithoutAllocatingMemoryIfProcessingHeaders) {
+  initialize();
+
+  MockRequestDecoder decoder;
+  EXPECT_CALL(callbacks_, newStream(_, _))
+      .WillOnce(Invoke([&](ResponseEncoder&, bool) -> RequestDecoder& { return decoder; }));
+
+  std::array<char, 1024> buffer;
+  OutputBufferStream ostream{buffer.data(), buffer.size()};
+
+  Buffer::OwnedImpl headers("POST / HTTP/1.1\r\n"
+                            "Host: host\r\n"
+                            "Accept-Language: en\r\n"
+                            "Connection: keep-alive\r\n"
+                            "Unfinished-Header: Not-Finished-Value");
+
+  auto status = codec_->dispatch(headers);
+  EXPECT_TRUE(status.ok());
+
+  // Dumps the header map without allocating memory
+  Stats::TestUtil::MemoryTest memory_test;
+  dynamic_cast<Http1::ServerConnectionImpl*>(codec_.get())->dumpState(ostream, 0);
+  EXPECT_EQ(memory_test.consumedBytes(), 0);
+
+  // Check dump contents for completed headers and partial headers.
+  EXPECT_THAT(
+      ostream.contents(),
+      testing::HasSubstr("absl::get<RequestHeaderMapPtr>(headers_or_trailers_): \n  ':authority', "
+                         "'host'\n  'accept-language', 'en'\n  'connection', 'keep-alive'"));
+  EXPECT_THAT(ostream.contents(),
+              testing::HasSubstr("header_parsing_state_: Value, current_header_field_: "
+                                 "Unfinished-Header, current_header_value_: Not-Finished-Value"));
+}
+
+TEST_F(Http1ServerConnectionImplTest, ShouldDumpDispatchBufferWithoutAllocatingMemory) {
+  initialize();
+
+  NiceMock<MockRequestDecoder> decoder;
+  EXPECT_CALL(callbacks_, newStream(_, _))
+      .WillOnce(Invoke([&](ResponseEncoder&, bool) -> RequestDecoder& { return decoder; }));
+
+  std::array<char, 1024> buffer;
+  OutputBufferStream ostream{buffer.data(), buffer.size()};
+
+  // Dump the body
+  Buffer::OwnedImpl request("POST / HTTP/1.1\r\n"
+                            "Content-Length: 11\r\n"
+                            "\r\n"
+                            "Hello Envoy");
```

Comment:
Would it be possible to split the request over 2 slices and verify that only the first slice is logged?

Something like:
```
Buffer::OwnedImpl request;
request.appendSliceForTest("POST / HTTP/1.1\r\n"
                            "Content-Length: 11\r\n"
                            "\r\n"
                            "Hello");
request.appendSliceForTest(" Envoy");

  EXPECT_THAT(ostream.contents(),
￼             HasSubstr("front_slice.length(): 44 ...
```

Reply:
Did something similar to this; We were using onMessageCompleteBase's callback from the parser to dump the data of the buffer. 

I added extraneous bytes in the other slice to  check that only the front slice is dumped. The content length is used to tell the parser the message should only have the front slice.